torch/csrc/jit/ir.cpp

```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+
+    for (; it_a != list_a.end() && it_b != list_b.end(); ++it_a, ++it_b) {
+      action(*it_a, *it_b);
+    }
+  }
+
+  std::ostream& printValueList(
+    std::ostream & out,
+    at::ArrayRef<const Value*> list
+  ) {
+    out << "(";
+    auto delimiter = "";
+    for (const auto* value : list) {
+      out << delimiter;
+      printValue(out, value);
+      delimiter = ", ";
+    }
+    out << ")";
+    return out;
+  }
+
+  void printAssignment(
+    std::ostream& out,
+    const Value* lhs,
+    const Value* rhs,
+    const size_t level
+  ) {
+    indent(out, level);
+    printValue(out, lhs);
+    out << " = ";
+    printValue(out, rhs);
+    out << std::endl;
+  }
+
+  std::ostream& printIf(
+    std::ostream & out,
+    const Node* node,
+    const size_t level
+  ) {
+    indent(out, level);
+    out << "if ";
+    const auto if_block = node->blocks()[0];
+    const auto else_block = node->blocks()[1];
+    printValue(out, node->inputs()[0]);
+    out << ":" << std::endl;
+
+    // Print node contents
+    printBlock(out, if_block, level + 1);
+
+    // Print if block output
+    dualIterator<const Value*>(
+      node->outputs(),
+      if_block->outputs(),
+      0, 0,
+      [&](const Value* node_output, const Value* return_input) {
+        printAssignment(out, node_output, return_input, level + 1);
+      }
+    );
+
+    indent(out, level);
+    out << "else:" << std::endl;
+    printBlock(out, else_block, level + 1);
+    dualIterator<const Value*>(
+      node->outputs(),
+      else_block->outputs(),
+      0, 0,
+      [&](const Value* node_output, const Value* return_input) {
+        printAssignment(out, node_output, return_input, level + 1);
+      }
+    );
+
+    return out;
+  }
+
+  bool isValueUsedLater(
+    const Value* val
+  ) const {
+    if (aliases_.find(val) != aliases_.end()) {
+      return true;
+    }
+    return val->uses().size() > 0;
+  }
+
+  std::ostream& printLoop(
+    std::ostream & out,
+    const Node* node,
+    const size_t level
+  ) {
+    const auto body_block = node->blocks()[0];
+    aliases_[body_block->inputs()[0]] = body_block->inputs()[0];
+
+    // Add temporaries for loop-carried dependencies
+    dualIterator<const Value*>(
+      node->inputs(),
+      body_block->inputs(),
+      1, 0,
+      [&](const Value* node_input, const Value* param_output) {
+        if (isValueUsedLater(param_output)) {
+          printAssignment(out, param_output, node_input, level);
+        }
+      }
+    );
+
+    dualIterator<const Value*>(
+      node->outputs(),
+      node->inputs(),
+      0, 2,
+      [&](const Value* node_output, const Value* return_input) {
+        if (isValueUsedLater(node_output)) {
+          printAssignment(out, node_output, return_input, level);
+        }
+      }
+    );
+
+    // Loop header
+    indent(out, level);
+    out << "while ";
+    printValue(out, body_block->inputs()[0]);
+    out << ":";
+    out << std::endl;
```

Comment:
nit: it's better to use `"\n"` instead of `std::endl`, because `endl` forces an IO buffer flush. We don't need it here, so better to avoid it.